demos/common/hfsm/az_iot_hfsm_sync_adapter.c

```diff
@@ -0,0 +1,309 @@
+static az_hfsm* active_hfsm = &provisioning_hfsm;
+#else
+static az_hfsm* active_hfsm = &hub_hfsm;
+#endif
+
+typedef enum
+{
+  AZ_IOT_HFSM_SYNC_DO_WORK = AZ_HFSM_EVENT(1),
+  AZ_IOT_HFSM_SYNC_DO_DELAY = AZ_HFSM_EVENT(2)
+} az_iot_hfsm_sync_event_type;
+
+static const az_hfsm_event az_hfsm_event_do_work = { AZ_IOT_HFSM_SYNC_DO_WORK, NULL };
+static const az_hfsm_event az_hfsm_event_do_delay = { AZ_IOT_HFSM_SYNC_DO_DELAY, NULL };
+
+static az_hfsm_event az_hfsm_sync_event_error = { AZ_IOT_ERROR, NULL };
+
+static int32_t delay_milliseconds;
+
+static int32_t azure_iot_sync(az_hfsm* me, az_hfsm_event event);
+static int32_t idle(az_hfsm* me, az_hfsm_event event);
+static int32_t running(az_hfsm* me, az_hfsm_event event);
+static int32_t timeout(az_hfsm* me, az_hfsm_event event);
+
+static az_hfsm_state_handler provisioning_and_hub_get_parent(az_hfsm_state_handler child_state)
+{
+  az_hfsm_state_handler parent_state;
+
+  if ((child_state == azure_iot_sync))
+  {
+    parent_state = NULL;
+  }
+  else 
+  {
+    parent_state = azure_iot_sync;
+  }
+
+  return parent_state;
+}
+
+static int32_t azure_iot_sync(az_hfsm* me, az_hfsm_event event)
+{
+  switch ((int32_t)event.type)
+  {
+    case AZ_HFSM_ENTRY:
+    case AZ_HFSM_EXIT:
+      break;
+    
+    default:
+      // Control should never reach this azure_iot_sync state.
+      LogInfo( ("azure_iot_sync: PANIC!") );
+      az_iot_hfsm_pal_critical_error(me);
+  }
+
+  return 0;
+}
+
+static int32_t idle(az_hfsm* me, az_hfsm_event event)
+{
+  int32_t ret = 0;
+
+  switch ((int32_t)event.type)
+  {
+    case AZ_HFSM_ENTRY:
+    case AZ_HFSM_EXIT:
+    case AZ_IOT_HFSM_SYNC_DO_WORK:
+      break;
+
+    case AZ_IOT_START:
+        LogInfo( ("idle: AZ_IOT_START") );
+
+        active_hfsm = me;
+        // TODO: remove internals access by moving to the start event data.
+        az_iot_hfsm_sync_adapter_pal_set_credentials( iot_hfsm._use_secondary_credentials );
+        ret = az_hfsm_transition_peer(me, idle, running);
+        break;
+        
+    default:
+      ret = AZ_HFSM_RET_HANDLE_BY_SUPERSTATE;
+  }
+
+  return ret;
+}
+
+static int32_t running(az_hfsm* me, az_hfsm_event event)
+{
+  int32_t ret = 0;
+
+  switch ((int32_t)event.type)
+  {
+    case AZ_HFSM_ENTRY:
+    case AZ_HFSM_EXIT:
+     break;
+
+    case AZ_IOT_HFSM_SYNC_DO_WORK:
+      LogInfo( ("running: AZ_IOT_HFSM_SYNC_DO_WORK") );
+      az_iot_hfsm_event_data_error status;
+
+#ifdef AZ_IOT_HFSM_PROVISIONING_ENABLED
+      if (me == &provisioning_hfsm)
+      {
+        status = az_iot_hfsm_sync_adapter_pal_run_provisioning();
+        if (status.type == AZ_IOT_OK)
+        {
+          ret = az_hfsm_post_event((az_hfsm*)(&iot_hfsm), az_hfsm_event_az_iot_provisioning_done);
+        }
+      }
+      else
+      {
+#endif
+        _az_PRECONDITION(me == &hub_hfsm);
+        
+        do
+        {
+            status = az_iot_hfsm_sync_adapter_pal_run_hub();
+        } while (status.type == AZ_IOT_OK);
```

Comment:
this will loop forever until error? That will trigger an error at line 185

Reply:
Yes - the flow keeps the device performing hub operations until an error is observed. 
The if below is unnecessary - will replace with a precondition instead.

Reply:
I take that back - there is a bug here.